docs(guides): update next.js tech guide to the relevant status

GeorgKrom opened the pull request.

illright commented on i18n/en/docusaurus-plugin-content-docs/current/guides/tech/with-nextjs.mdx:
```diff
@@ -1,72 +1,73 @@
+## Conflict between FSD and Next.js in the `app` layer
 
-## Conflict between FSD and NextJS on `pages` layer {#pages-conflict}
+Next.js suggests using the `app` folder to define application routes. Next.js expects files in the `app` folder to correspond to URLs.
```
I see that it's not you who introduced these issues, but thought that now's a good time to fix them anyway

illright commented on i18n/en/docusaurus-plugin-content-docs/current/guides/tech/with-nextjs.mdx:
```diff
@@ -1,72 +1,73 @@
-NextJS suggests using the `pages` folder to define application routes. NextJS expects files in the `pages` folder to match URLs.
-This routing mechanism **does not correspond** to the FSD concept, since it is not possible to maintain a flat slice structure in such a routing mechanism.
+You should also add a `pages` folder to the project root because App Router is compatible with Pages Router.
+The best way is to put an empty `.gitkeep` or `README.md` file inside describing the need to save this folder.
```
suggestion (non-blocking): I would focus the recommendation to the README file instead of `.gitkeep` as it's not obvious why this empty folder is needed, so project docs could really come in handy later down the line

illright commented on i18n/en/docusaurus-plugin-content-docs/current/guides/tech/with-nextjs.mdx:
```diff
@@ -1,72 +1,73 @@
-├── shared
-├── widgets
+### Route Handlers
+It is suggested to use a new `routes` segment in the `app` or `pages` layer to work with Route Handlers.
```
issue: this part isn't clear to me. You mention that it's recommended to use the `routes` segment, but the code block below doesn't demonstrate it.

suggestion: let's also add a code block where `getExample` is defined so that it's clear how and where to define route handlers

But also, should this really be in the FSD structure? I feel like mixing in unrelated backend code into the frontend code folder might make it hard to find. Let's discuss and refine this point a bit, perhaps also get people from the chat to give their opinion

illright replied:
В целом, я согласен с тобой, что есть польза в том, чтоб описать, куда мы предлагаем класть простенькие API-роуты, раз уж мы всю структуру перерыли. Давай может тогда вот так поступим:
- уберем из рекомендаций предложение хранить их в pages — они не имеют отношения к страницам, пусть уж будут в `app` только
- сделаем два код-блока в этой секции, первый в `src/app/api-routes` (мне больше нравится api-routes, потому что тогда не путаем с роутами, по которым находятся страницы), и в нем прямо полноценное описание функции-хендлера `getExampleData`. Внутри этой функции можем даже заюзать `db` из `shared/db`, чтоб продемонстрировать реюз коннекшна между роут-хендлерами и серверными компонентами
- второй код-блок в `app/api/example`, и он реэкспортирует `getExampleData as GET`

illright commented on i18n/en/docusaurus-plugin-content-docs/current/guides/tech/with-nextjs.mdx:
```diff
@@ -1,72 +1,73 @@
 
-Keep in mind that it's highly recommended to document this rename prominently in your project's README or internal documentation. This rename is a part of your ["project knowledge"][project-knowledge].
+### Additional recommendations
+- Use the `queries` segment in the `shared` layer to describe database queries and their further use in higher layers.
```
issue: I think that `queries` here describes what the code is, not what the code is for (see https://github.com/feature-sliced/steiger/blob/master/packages/steiger-plugin-fsd/src/segments-by-purpose/README.md)

suggestion: I prefer to keep the logic of server actions and data loaders in the `api` segment of pages that need it, but let's discuss this in more detail too

illright commented on i18n/en/docusaurus-plugin-content-docs/current/guides/tech/with-nextjs.mdx:
```diff
@@ -75,47 +76,16 @@ const App = ({ Component, pageProps }: AppProps) => {
+## See also
 
-[project-knowledge]: /docs/about/understanding/knowledge-types
-[ext-app-router-stackblitz]: https://stackblitz.com/edit/stackblitz-starters-aiez55?file=README.md
```
suggestion: let's not remove this example, I think it's useful for people to see a functioning app that implements this recommendation. If there's something you'd like to change about it — let me know, I have write access to it

illright commented on i18n/en/docusaurus-plugin-content-docs/current/guides/tech/with-nextjs.mdx:
```diff
@@ -1,72 +1,73 @@
+### Importing functionality into `pages/_app.tsx`
 
-To solve the problem of missing the `app` folder in the NextJS structure, you can create an `App` component inside the `app` layer and import the `App` component into `pages/_app.tsx` so that NextJS can work with it.
+To solve the problem of missing the `app` folder in the Next.js structure, you can create an `App` component inside the `app` layer and import it into `pages/_app.tsx` so that Next.js can work with it.
```
issue: to be honest, this section doesn't really make sense, I don't remember why it was added

suggestion: maybe we should rewrite it with proper guidelines for the Pages router?

GeorgKrom replied:
Yep, I will rewrite it

illright commented:
Привет, спасибо, да, завтра посмотрю основательно апдейты

illright commented on i18n/en/docusaurus-plugin-content-docs/current/guides/tech/with-nextjs.mdx:
````diff
@@ -1,121 +1,118 @@
-```tsx
-// pages/_app.tsx
+### Additional recommendations
+- Use the `api` segment in the `shared` layer to describe database queries and their further use in higher layers.
````
issue: Пока писал прошлый коммент, понял, что мне все-таки не очень нравится идея описывать какие-то переиспользуемые запросы к БД в сегменте `api`, это все-таки совсем другое. 

suggestion: Как насчет порекомендовать тут тоже использовать сегмент `db`? Будет тогда следующая параллель:
- в корне сегмента shared/api обычно описывается какой-то API-клиент, который используется другими, более конкретными API-функциями в shared/api/<endpoint> (сгруппированных по эндпоинту)
- в корне сегмента shared/db будет описываться клиент БД, будь то дриззл или призма или еще че, и он затем будет использоваться более конкретными функциями-запросами, сгруппированными по основной таблице или бизнес-сущности

illright replied:
Тогда и следующую строчку тоже надо будет обновить

illright commented on i18n/en/docusaurus-plugin-content-docs/current/guides/tech/with-nextjs.mdx:
```diff
@@ -1,121 +1,118 @@
 
-import 'app/styles/index.scss'
+## Legacy approach with Pages Router
+Page routers should be placed in the `pages` folder in the root of the project, similar to `app` for the App Router.
```
suggestion (non-blocking):
```suggestion
When using the Pages Router, routes should be placed in the `pages` folder in the root of the project, similar to `app` for the App Router.
```

illright commented on i18n/en/docusaurus-plugin-content-docs/current/guides/tech/with-nextjs.mdx:
````diff
@@ -1,121 +1,118 @@
 ```
 
-## Dealing with App Router {#app-router}
+### Functionality re-export from the public API `src/app/index.ts` to `pages/_app.tsx`
````
issue: это какая-то не очень понятная секция, без текста, только кодовый блок. Помимо этого, мне кажется, Steiger будет ругаться на индекс-файл в app, но я не уверен

suggestion: хочется найти какое-то более стабильное место кастомному компоненту App. При этом, не хочется выдумывать для него другие имена, которые не будут знакомы Next-разработчикам. Может, забить и просто предложить людям создать в слое `app` сегмент  `_app`? Повторение, конечно, но зато сразу понятно всем, кто знают Pages Router. Альтернативно можно предложить `custom-app`

И давай заголовок этой секции сделаем вот такой:

```suggestion
### Custom `_app` component
```

И чуть текста добавим, типа "You can place your Custom App component in `src/app/_app` or `src/app/custom-app`, и затем ниже привести опять же два код-блока — один с содержимым кастомного аппа, другой с реэкспортом в нужное место

illright commented on i18n/en/docusaurus-plugin-content-docs/current/guides/tech/with-nextjs.mdx:
````diff
@@ -1,121 +1,118 @@
 ```
 
-### Renaming the `pages` layer within the FSD structure
+### Example of re-exporting a page from `src/pages/example` to `app/example/page.tsx`
````
issue: тут тоже раздел получается немного неописательным, и на мой взгляд, незаслуженно отделенным от прошлого раздела.

suggestion: мб не делать эти примеры заголовками, а просто текстом обычным оставить?
```suggestion
Example of re-exporting a page from `src/pages/example` to `app/example/page.tsx`:
```

illright reviewed (changes requested):
Так, наконец-то собрался с мыслями. Спасибо за терпение :) Отписал свои мысли по поводу вопросов, которые поднимал на прошлом раунде комментариев

GeorgKrom commented:
> Thanks for addressing the comments! I have some more, but we're very close :)
> 
> P.S. sorry for switching between English and Russian randomly lol, I just wanna make sure that this conversation could be at least roughly followed by non-Russian-speaking contributors who might stumble upon this PR. Feel free to continue writing in Russian btw

aahh, now i get the point XD

illright commented on i18n/en/docusaurus-plugin-content-docs/current/guides/tech/with-nextjs.mdx:
```diff
@@ -1,121 +1,153 @@
-which plays the role of a wrapping component for all project pages.
+### Route Handlers
+It is suggested to use a new `api-routes` segment in the `app` layer to work with Route Handlers.
+However, you should avoid overusing them due to code maintenance complexity and violation of separation of concerns principles.
```
could you also update that in the Russian version please?

illright commented:
By the way, I also remembered that someone complained that we didn't write out the reason for creating an empty `pages` folder clearly enough, could you also expand on that, for example, say that adding an empty folder will stop Next from treating `src/pages` as the Pages Router folder?

https://github.com/feature-sliced/documentation/issues/796

GeorgKrom commented:
> By the way, I also remembered that someone complained that we didn't write out the reason for creating an empty `pages` folder clearly enough, could you also expand on that, for example, say that adding an empty folder will stop Next from treating `src/pages` as the Pages Router folder?
> 
> #796

done
